File: Windows/LineScope.Core/ColorSpaces.cs
namespace LineScope.Core;

public enum ColorModel { Rgba, Hsl, Cmyk }

public static class ColorModelInfo
{
    public static readonly ColorModel[] All = [ColorModel.Rgba, ColorModel.Hsl, ColorModel.Cmyk];

    public static string DisplayName(this ColorModel m) => m switch
    {
        ColorModel.Rgba => "RGBA",
        ColorModel.Hsl => "HSL",
        _ => "CMYK",
    };

    public static string[] Channels(this ColorModel m) => m switch
    {
        ColorModel.Rgba => ["Red", "Green", "Blue", "Alpha"],
        ColorModel.Hsl => ["Hue", "Saturation", "Lightness"],
        _ => ["Cyan", "Magenta", "Yellow", "Black"],
    };

    /// <summary>Plot range for a channel; hue is in degrees, everything else in 0…1.</summary>
    public static (double Min, double Max) RangeOfChannel(this ColorModel m, int channel) =>
        m == ColorModel.Hsl && channel == 0 ? (0, 360) : (0, 1);
}

public static class ColorConversion
{
    // HSL (hue in degrees)

    public static (float H, float S, float L) Hsl(Rgba c)
    {
        float maxV = Math.Max(c.R, Math.Max(c.G, c.B));
        float minV = Math.Min(c.R, Math.Min(c.G, c.B));
        float l = (maxV + minV) / 2;
        float d = maxV - minV;
        if (!(d > 1e-7f)) return (0, 0, l);
        float s = d / (1 - MathF.Abs(2 * l - 1));
        float h;
        if (maxV == c.R) h = ((c.G - c.B) / d) % 6;
        else if (maxV == c.G) h = (c.B - c.R) / d + 2;
        else h = (c.R - c.G) / d + 4;
        h *= 60;
        if (h < 0) h += 360;
        return (h, Math.Clamp(s, 0, 1), l);
    }

    public static Rgba FromHsl(float h, float s, float l, float alpha = 1)
    {
        float c = (1 - MathF.Abs(2 * l - 1)) * s;
        float hp = h / 60;
        float x = c * (1 - MathF.Abs(hp % 2 - 1));
        (float r, float g, float b) = hp switch
        {
            < 1 => (c, x, 0f),
            < 2 => (x, c, 0f),
            < 3 => (0f, c, x),
            < 4 => (0f, x, c),
            < 5 => (x, 0f, c),
            _ => (c, 0f, x),
        };
        float m = l - c / 2;
        return new Rgba(r + m, g + m, b + m, alpha);
    }

    // CMYK (naive device-independent conversion, K = 1 - max(R,G,B))

    public static (float C, float M, float Y, float K) Cmyk(Rgba c)
    {
        float k = 1 - Math.Max(c.R, Math.Max(c.G, c.B));
        if (!(k < 1 - 1e-7f)) return (0, 0, 0, 1);
        float d = 1 - k;
        return ((1 - c.R - k) / d, (1 - c.G - k) / d, (1 - c.B - k) / d, k);
    }

    public static Rgba FromCmyk(float c, float m, float y, float k, float alpha = 1) =>
        new((1 - c) * (1 - k), (1 - m) * (1 - k), (1 - y) * (1 - k), alpha);

    // Channel extraction

    public static float Value(Rgba color, ColorModel model, int channel)
    {
        switch (model)
        {
            case ColorModel.Rgba:
                return channel switch { 0 => color.R, 1 => color.G, 2 => color.B, _ => color.A };
            case ColorModel.Hsl:
                var v = Hsl(color);
                return channel switch { 0 => v.H, 1 => v.S, _ => v.L };
            default:
                var k = Cmyk(color);
                return channel switch { 0 => k.C, 1 => k.M, 2 => k.Y, _ => k.K };
        }
    }
}

File: Packages/LineScopeCore/Sources/LineScopeCore/ColorSpaces.swift
import Foundation

public enum ColorModel: String, CaseIterable, Identifiable, Sendable {
    case rgba
    case hsl
    case cmyk

    public var id: String { rawValue }

    public var displayName: String {
        switch self {
        case .rgba: "RGBA"
        case .hsl: "HSL"
        case .cmyk: "CMYK"
        }
    }

    public var channels: [String] {
        switch self {
        case .rgba: ["Red", "Green", "Blue", "Alpha"]
        case .hsl: ["Hue", "Saturation", "Lightness"]
        case .cmyk: ["Cyan", "Magenta", "Yellow", "Black"]
        }
    }

    /// Plot range for a channel; hue is in degrees, everything else in 0...1.
    public func range(ofChannel channel: Int) -> ClosedRange<Double> {
        self == .hsl && channel == 0 ? 0...360 : 0...1
    }
}

public enum ColorConversion {
    // MARK: HSL (hue in degrees)

    public static func hsl(from c: RGBA) -> (h: Float, s: Float, l: Float) {
        let maxV = max(c.r, c.g, c.b)
        let minV = min(c.r, c.g, c.b)
        let l = (maxV + minV) / 2
        let d = maxV - minV
        guard d > 1e-7 else { return (0, 0, l) }
        let s = d / (1 - abs(2 * l - 1))
        var h: Float
        if maxV == c.r {
            h = ((c.g - c.b) / d).truncatingRemainder(dividingBy: 6)
        } else if maxV == c.g {
            h = (c.b - c.r) / d + 2
        } else {
            h = (c.r - c.g) / d + 4
        }
        h *= 60
        if h < 0 { h += 360 }
        return (h, min(max(s, 0), 1), l)
    }

    public static func rgb(h: Float, s: Float, l: Float, alpha: Float = 1) -> RGBA {
        let c = (1 - abs(2 * l - 1)) * s
        let hp = h / 60
        let x = c * (1 - abs(hp.truncatingRemainder(dividingBy: 2) - 1))
        let (r, g, b): (Float, Float, Float)
        switch hp {
        case ..<1: (r, g, b) = (c, x, 0)
        case ..<2: (r, g, b) = (x, c, 0)
        case ..<3: (r, g, b) = (0, c, x)
        case ..<4: (r, g, b) = (0, x, c)
        case ..<5: (r, g, b) = (x, 0, c)
        default: (r, g, b) = (c, 0, x)
        }
        let m = l - c / 2
        return RGBA(r: r + m, g: g + m, b: b + m, a: alpha)
    }

    // MARK: CMYK (naive device-independent conversion, K = 1 - max(R,G,B))

    public static func cmyk(from c: RGBA) -> (c: Float, m: Float, y: Float, k: Float) {
        let k = 1 - max(c.r, c.g, c.b)
        guard k < 1 - 1e-7 else { return (0, 0, 0, 1) }
        let d = 1 - k
        return ((1 - c.r - k) / d, (1 - c.g - k) / d, (1 - c.b - k) / d, k)
    }

    public static func rgb(c: Float, m: Float, y: Float, k: Float, alpha: Float = 1) -> RGBA {
        RGBA(r: (1 - c) * (1 - k), g: (1 - m) * (1 - k), b: (1 - y) * (1 - k), a: alpha)
    }

    // MARK: Channel extraction

    public static func value(of color: RGBA, model: ColorModel, channel: Int) -> Float {
        switch model {
        case .rgba:
            return [color.r, color.g, color.b, color.a][channel]
        case .hsl:
            let v = hsl(from: color)
            return [v.h, v.s, v.l][channel]
        case .cmyk:
            let v = cmyk(from: color)
            return [v.c, v.m, v.y, v.k][channel]
        }
    }
}
